attest/README.md

```diff
@@ -1,3 +1,108 @@
+
+## Trusted vs. Untrusted
+
+So how do you do this? Let's start by splitting application code into two pieces: the normal OS application, which you execute like any other process from a shell or launcher; and the enclave, which is a shared library which has one or more functions which can be called from outside the enclave (it is very similar to a `dlopen()`-style plugin library, and when built in "simulation" mode, that's literally what the SGX SDK does).
+
+Intel calls the normal application the ***untrusted*** code, and the enclave library the ***trusted*** code, and in the MobileCoin universe these are literally separate code compiled from independent `cargo build` runs. Additionally, only specifically marked functions in untrusted can be called by trusted code, and visa-versa. This means trusted code cannot perform any IO directly, it can only ask untrusted to act on its behalf.
+
+OK, now we have our code split between outside-the-enclave (untrusted) and inside-the-enclave (trusted). As parenthetically noted above, Intel does have a "simulated" enclave, where they just take the "trusted" code and turn it into a plugin module for your application---how can you tell the difference between a "simulated" enclave and a real one if they're running the same code?
+
+## Attestation
+
+This "telling the difference" is handled by attestation, which is a multi-layered technique using building blocks from lower levels.
+
+1. Local attestation (most of the actual work of attestation is done locally, on the host)
+1. Quoting enclaves (by passing data to the quoting enclave)
+1. IAS API call (a REST API call to IAS to sign off on the local attestation data)
+1. Verification (verifying the result from IAS is what we want)
+
+### Between Enclaves on the same CPU
+
+> Don't skip this section just because you're interested in remote attestation. It's a core building block if you want to understand how the system works.
+
+The first thing to know about attestation is that it begins with Alice The Enclave asking the CPU for it's own `EREPORT`. The CPU will fill in a data structure containing information about Alice.
+
+The contents of Alice's `EREPORT` can be used to construct a `TARGETINFO` data structure, which is given to Bob The Enclave. Bob uses Alice's targetting information when he calls `EREPORT` from within his context, and he CPU fills in the report as before, and adds an authentication code so Alice can verify Bob's report is valid.
+
+Bob sends his `EREPORT` (with authentication code) to Alice, and she verifies the code using information available only to her and the CPU, and she can trust the report that she got from Bob is actually from Bob. That is, she's authenticated the report, and knows there was a Bob enclave out there, and she knows she has a report for it.
```

Comment:
I agree with EREPORT (the command) vs. REPORT (the structure), but I'm going to keep it simple, there's no need to discuss EGETKEY (which is never directly called in any of our code).

Reply:
Ok, but I think the 'message' concept is important. Attestation is meaningless without the report message.

Reply:
I disagree, not because `REPORTDATA` isn't the message (in a cryptographic sense), but because it should not be treated as a "message" by an author of an application that wants to use an enclave. It may be interesting to know, but it isn't necessary to understanding, and will lead folks down the wrong path.

For example: I think one mistake we made was putting the block signing key into the AVR. Now, if we want to save round-trips by adding more details to authentication (e.g. adding the configured minimum fee to that would save a round-trip during tx construction, and an enclave verifier could check that the fees match directly, and provide better error reporting, rather than stuffing it into the noise prologue), we're kind of screwed. If we'd treated those bytes as an authentication tag for a larger data structure instead (i.e. as a bootstrap for a *real* message), I think we'd be in a better spot.

Reply:
I don't really understand. Isn't proving a piece of data (message) was made by an enclave the entire raison d'etre of attestation? You don't gain anything from a `REPORT` or AVR that has no message.